### PR TITLE
Evaluate Aggregate Assignment Only When FILTER Clause Is True

### PR DESCRIPTION
**Pull request by atris**
FILTER clauses in aggregates are today planned as mask field in aggregation.
This implies that we add a projection underneath the aggregate which will
project the aggregate expression and mask expression. However, this leads
to evaluating the aggregate expression assignment for all the input tuples,
whereas the assignment should be evaluated only for tuples where the filter
clause is true.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 1]
```diff
@@ -351,6 +355,88 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
...
+            List<FilterExpressionWithAggregate> filterExpressionWithAggregates,
+            Iterable<Expression> groupAndSortArguments,
+            ImmutableList.Builder argumentsAfterFilter)
+    {
+        TranslationMap outputTranslations = new TranslationMap(subPlan.getRelationPlan(), analysis, lambdaDeclarationToSymbolMap);
+
+        Assignments.Builder projections = Assignments.builder();
+        Assignments.Builder filterArguments = Assignments.builder();
+        
+        for (FilterExpressionWithAggregate currentFilterExpressionWithAggregate : filterExpressionWithAggregates) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: s/currentFilterExpressionWithAggregate/filterExpressionWithAggregate

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 2]
```diff
@@ -351,6 +355,88 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
                 analysis.getParameters());
     }
 
+    private PlanBuilder project(
+            PlanBuilder subPlan,
+            List<FilterExpressionWithAggregate> filterExpressionWithAggregates,
+            Iterable<Expression> groupAndSortArguments,
+            ImmutableList.Builder argumentsAfterFilter)
+    {
+        TranslationMap outputTranslations = new TranslationMap(subPlan.getRelationPlan(), analysis, lambdaDeclarationToSymbolMap);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: s/outputTranslations/translations

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 3]
```diff
@@ -419,6 +505,21 @@ private PlanBuilder explicitCoercionSymbols(PlanBuilder subPlan, Iterable<Symbol
                 analysis.getParameters());
     }
 
+    private ImmutableList<FilterExpressionWithAggregate> getFilterExpressions(QuerySpecification node)
+    {
+        ImmutableList.Builder resultBuilder = ImmutableList.builder();
+        analysis.getAggregates(node).stream().forEach(currentAggregate -> {
+            FilterExpressionWithAggregate filterExpressionWithAggregate = new FilterExpressionWithAggregate(
+                    currentAggregate.getFilter().isPresent() ? currentAggregate.getFilter().get() : null,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: Optional is better than plain `null`

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: Why not use `FunctionCall` instead of `FilterExpressionWithAggregate` as a parameter to your `project` method?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 4]
```diff
@@ -419,6 +505,21 @@ private PlanBuilder explicitCoercionSymbols(PlanBuilder subPlan, Iterable<Symbol
...
+                    currentAggregate,
+                    currentAggregate.getArguments());
+
+            resultBuilder.add(filterExpressionWithAggregate);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: use `stream().map(..).collect(toImmutableList())`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 5]
```diff
@@ -419,6 +505,21 @@ private PlanBuilder explicitCoercionSymbols(PlanBuilder subPlan, Iterable<Symbol
                 analysis.getParameters());
     }
 
+    private ImmutableList<FilterExpressionWithAggregate> getFilterExpressions(QuerySpecification node)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: return `List`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 6]
```diff
@@ -951,4 +1080,45 @@ public static SortOrder toSortOrder(SortItem sortItem)
             }
         }
     }
+
+    private class FilterExpressionWithAggregate
```
[the remark this answers is not in the document]
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: There is no clean way of mapping a ``FunctionCall``  to a list of arguments which are rewritten to ``CASE`` expressions if that is needed, hence ``FilterExpressionWithAggregate``  is needed. While the class exists, I thought that piggybacking and delegating some of the filter processing to ``FilterExpressionWithAggregate`` and letting the consumers only deal with final results will be worth it, but I can change ``FilterExpressionWithAggregate`` to only be a mapping between final arguments and ``FunctionCall``.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 7]
```diff
@@ -351,6 +355,88 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
                 analysis.getParameters());
     }
 
+    private PlanBuilder project(
```
[the remark this answers is not in the document]
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: I put it with the remaining ``project`` methods, and named it because it is adding projections for the aggregate arguments. Does that make sense, please?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestAggregations.java [hunk 8]
```diff
@@ -1260,7 +1260,7 @@ public void testOrderedAggregations()
                         "WHERE orderkey IN (1, 2, 3, 4, 5) " +
                         "GROUP BY GROUPING SETS ((), (orderpriority), (orderpriority, custkey))",
                 "VALUES " +
-                        "(NULL, NULL , ('F', 'O', 'O'))," +
+                        "(NULL, NULL , ('O', 'F', 'O'))," +
```
[the remark this answers is not in the document]
Comment (atris) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestAggregations.java]: Yes, I noticed it today morning in another test and pushed the fixes and then saw your comment :) Thanks for pointing this out

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 9]
```diff
@@ -86,4 +86,29 @@ public void testGroupingSets()
                         "(2, BIGINT '4', BIGINT '1'), " +
                         "(CAST(NULL AS INTEGER), BIGINT '5', BIGINT '2')");
     }
+
```
[the remark this answers is not in the document]
Comment (atris) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: With the PR

``presto:sf1> explain select sum(a + 1) filter (where a + 1 > 0) from (values (1), (2), (3)) t(a);
                                             Query Plan                         
--------------------------------------------------------------------------------
 - Output[_col0] => [sum:bigint]                                                
         Cost: {rows: 1 (9B), cpu: 111.00, memory: 18.00, network: 0.00}        
         _col0 := sum                                                           
     - Aggregate(FINAL) => [sum:bigint]                                         
             Cost: {rows: 1 (9B), cpu: 111.00, memory: 18.00, network: 0.00}    
             sum := "sum"("sum_9")                                              
         - LocalExchange[SINGLE] () => sum_9:bigint                             
                 Cost: {rows: 1 (9B), cpu: 102.00, memory: 9.00, network: 0.00} 
             - Aggregate(PARTIAL) => [sum_9:bigint]                             
                     Cost: {rows: 1 (9B), cpu: 102.00, memory: 9.00, network: 0.
                     sum_9 := "sum"("expr_3") (mask = expr_2)                   
                 - Project[] => [expr_3:bigint, expr_2:boolean]                 
                         Cost: {rows: 3 (33B), cpu: 69.00, memory: 0.00, network
                         expr_3 := (CASE "expr_2" WHEN true THEN CAST(("field" +
                     - Project[] => [expr_2:boolean, field:integer]             
                             Cost: {rows: 3 (21B), cpu: 36.00, memory: 0.00, net
                             expr_2 := (("field" + 1) > 0)                      
                         - LocalExchange[ROUND_ROBIN] () => field:integer       
                                 Cost: {rows: 3 (15B), cpu: 15.00, memory: 0.00,
                             - Values => [field:integer]                        
                                     Cost: {rows: 3 (15B), cpu: 0.00, memory: 0.
                                     (1)                                        
                                     (2)                                        
                                     (3) ``



Without the PR

``presto:sf1> explain select sum(a + 1) filter (where a + 1 > 0) from (values (1), (2), (3)) t(a);
                                          Query Plan                            
--------------------------------------------------------------------------------
 - Output[_col0] => [sum:bigint]                                                
         Cost: {rows: 1 (10B), cpu: 160.00, memory: 20.00, network: 0.00}       
         _col0 := sum                                                           
     - Aggregate(FINAL) => [sum:bigint]                                         
             Cost: {rows: 1 (10B), cpu: 160.00, memory: 20.00, network: 0.00}   
             sum := "sum"("sum_9")                                              
         - LocalExchange[SINGLE] () => sum_9:bigint                             
                 Cost: {rows: 1 (10B), cpu: 150.00, memory: 10.00, network: 0.00
             - Aggregate(PARTIAL) => [sum_9:bigint]                             
                     Cost: {rows: 1 (10B), cpu: 150.00, memory: 10.00, network: 
                     sum_9 := "sum"("expr_1") (mask = expr_2)                   
                 - Project[] => [expr_1:bigint, expr_2:boolean]                 
                         Cost: {rows: 3 (60B), cpu: 90.00, memory: 0.00, network
                         expr_1 := CAST(("field" + 1) AS bigint)                
                         expr_2 := (("field" + 1) > 0)                          
                     - LocalExchange[ROUND_ROBIN] () => field:integer           
                             Cost: {rows: 3 (30B), cpu: 30.00, memory: 0.00, net
                         - Values => [field:integer]                            
                                 Cost: {rows: 3 (30B), cpu: 0.00, memory: 0.00, 
                                 (1)                                            
                                 (2)                                            
                                 (3)   ``

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 10]
```diff
@@ -86,4 +86,29 @@ public void testGroupingSets()
...
+                        "        VALUES\n" +
+                        "            ('1', 'a', 'good'),\n" +
+                        "            ('2', 'b', 'good'),\n" +
+                        "            ('x', 'c', 'bad')\n" +
+                        "    ) AS t (v, k, name)\n" +
+                        ")\n" +
+                        "SELECT\n" +
+                        "    ARRAY_AGG(CAST(v AS BIGINT)) FILTER (WHERE name = 'good') AS col2\n" +
+                        "FROM test",
+                "VALUES (ARRAY[BIGINT '1', BIGINT '2'])");
```
[the remark this answers is not in the document]
Comment (atris) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: No, it will fail when trying to cast ``x`` as ``BIGINT``

**Comment by atris** [comment]
Fixes #11205

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 11]
```diff
@@ -351,6 +355,87 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
...
+            PlanBuilder subPlan,
+            ImmutableList<FilterExpressionWithAggregate> filterExpressionWithAggregates,
+            ImmutableList<Expression> groupAndSortArguments,
+            ImmutableList.Builder argumentsAfterFilter)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: * Use `List` here, not `ImmutableList` nor `ImmutableList.Builder`.
* `ImmutableList.Builder argumentsAfterFilter` you are missing type here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: `getFilterExpressions` can return `List`

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: Nothing changes. `expr_3` and `expr_2` are evaluationing `x + 1`. It could due a cast. Can you please test:
```
SELECT sum(x + 1) FILTER (WHERE x + 1 > 0) FROM VALUES (BIGINT '1', BIGINT '2', BIGINT '3');
```

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 12]
```diff
@@ -86,4 +86,29 @@ public void testGroupingSets()
...
+    @Test
+    public void testFilterProjectionExecutedConditionallyInCTE()
+    {
+        assertions.assertQuery(
+                "WITH test AS (" +
+                        "    SELECT * FROM (" +
+                        "        VALUES" +
+                        "            ('1', 'a', 'good'),\n" +
+                        "            ('2', 'b', 'good'),\n" +
+                        "            ('x', 'c', 'bad')\n" +
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: please remove all `\n`

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 13]
```diff
@@ -86,4 +86,29 @@ public void testGroupingSets()
                         "(2, BIGINT '4', BIGINT '1'), " +
                         "(CAST(NULL AS INTEGER), BIGINT '5', BIGINT '2')");
     }
+
+    @Test
+    public void testFilterProjectionExecutedConditionally()
+    {
+        assertions.assertQuery("select sum(1 / a) filter (where a <> 0) from (values (1), (0)) t(a)",
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: put each argument in separate line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: > I can change FilterExpressionWithAggregate to only be a mapping between final arguments and FunctionCall.

Sounds good to me.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (atris) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: ``presto:sf1> explain SELECT sum(x + 1) FILTER (WHERE x + 1 > 0) FROM (VALUES (BIGINT '1'), (BIGINT '2'), (BIGINT '3')) t(x);
                                            Query Plan                          
--------------------------------------------------------------------------------
 - Output[_col0] => [sum:bigint]                                                
         Cost: {rows: 1 (9B), cpu: 135.00, memory: 18.00, network: 0.00}        
         _col0 := sum                                                           
     - Aggregate(FINAL) => [sum:bigint]                                         
             Cost: {rows: 1 (9B), cpu: 135.00, memory: 18.00, network: 0.00}    
             sum := "sum"("sum_9")                                              
         - LocalExchange[SINGLE] () => sum_9:bigint                             
                 Cost: {rows: 1 (9B), cpu: 126.00, memory: 9.00, network: 0.00} 
             - Aggregate(PARTIAL) => [sum_9:bigint]                             
                     Cost: {rows: 1 (9B), cpu: 126.00, memory: 9.00, network: 0.
                     sum_9 := "sum"("expr_3") (mask = expr_2)                   
                 - Project[] => [expr_3:bigint, expr_2:boolean]                 
                         Cost: {rows: 3 (33B), cpu: 93.00, memory: 0.00, network
                         expr_3 := (CASE "expr_2" WHEN true THEN ("field" + BIGI
                     - Project[] => [expr_2:boolean, field:bigint]              
                             Cost: {rows: 3 (33B), cpu: 60.00, memory: 0.00, net
                             expr_2 := (("field" + BIGINT '1') > BIGINT '0')    
                         - LocalExchange[ROUND_ROBIN] () => field:bigint        
                                 Cost: {rows: 3 (27B), cpu: 27.00, memory: 0.00,
                             - Values => [field:bigint]                         
                                     Cost: {rows: 3 (27B), cpu: 0.00, memory: 0.
                                     (BIGINT '1')                               
                                     (BIGINT '2')                               
                                     (BIGINT '3')     ``

**Comment by atris** [comment]
I will squash commits post review and before commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 14]
```diff
@@ -351,6 +356,92 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
...
     }
 
+    private PlanBuilder project(
+            PlanBuilder subPlan,
+            List<FunctionCall> aggregates,
+            ImmutableList<Expression> groupAndSortArguments,
+            List<Expression> argumentsAfterFilter,
+            List<FunctionCallWithFinalArguments> functionCallToFinalArguments)
+    {
+        TranslationMap translations = new TranslationMap(subPlan.getRelationPlan(), analysis, lambdaDeclarationToSymbolMap);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: I would divide this method into two. First I would handle the projection related to filter predicates and then projection with arguments. The way you wrote it is very mixed up, hard to follow the code.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 15]
```diff
@@ -428,42 +447,71 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
...
+                .collect(toImmutableList());
+
+        Iterable<Expression> inputs = Iterables.concat(groupByExpressions, aggregateArguments, sortArguments, filterArguments);
+        ImmutableList<Expression> arguments = ImmutableList.copyOf(inputs);
 
-        Iterable<Expression> inputs = Iterables.concat(groupByExpressions, arguments.build());
         subPlan = handleSubqueries(subPlan, node, inputs);
 
-        if (!Iterables.isEmpty(inputs)) { // avoid an empty projection if the only aggregation is COUNT (which has no arguments)
+        Optional<List<FunctionCallWithFinalArguments>> functionCallWithFinalArgumentsOptional = Optional.empty();
```
[the remark this answers is not in the document]
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: As discussed offline, rewriting the aggregate to ``CASE`` expression earlier than the place where it is done today will lead to losing original aggregates, and since ``aggregate`` will return ``PlanBuilder`` only with final translation, the ``project`` method called after the ``aggregate`` will project symbols from ``analysis``'s output expression, so there will be a dependency mismatch.
